Patch-Source: https://git.deuxfleurs.fr/Deuxfleurs/garage/pulls/372
--
From d7a3de8a7b20723c1aba3949f7781d33a79ad943 Mon Sep 17 00:00:00 2001
From: Jakub Jirutka <jakub@jirutka.cz>
Date: Sat, 3 Sep 2022 23:40:44 +0200
Subject: [PATCH 2/3] Make OTLP exporter optional via feature "telemetry-otlp"

opentelemetry-otlp add 48 (!) extra dependencies and increases the
size of the garage binary by ~11 % (with fat LTO).

diff --git a/src/api/Cargo.toml b/src/api/Cargo.toml
--- a/src/api/Cargo.toml
+++ b/src/api/Cargo.toml
@@ -55,8 +55,9 @@
 
 opentelemetry = "0.17"
 opentelemetry-prometheus = "0.10"
-opentelemetry-otlp = "0.10"
+opentelemetry-otlp = { version = "0.10", optional = true }
 prometheus = "0.13"
 
 [features]
 k2v = [ "garage_util/k2v", "garage_model/k2v" ]
+telemetry-otlp = ["opentelemetry-otlp"]
diff --git a/src/garage/Cargo.toml b/src/garage/Cargo.toml
--- a/src/garage/Cargo.toml
+++ b/src/garage/Cargo.toml
@@ -54,7 +54,7 @@
 
 opentelemetry = { version = "0.17", features = [ "rt-tokio" ] }
 opentelemetry-prometheus = "0.10"
-opentelemetry-otlp = "0.10"
+opentelemetry-otlp = { version = "0.10", optional = true }
 prometheus = "0.13"
 
 [dev-dependencies]
@@ -74,3 +74,5 @@
 [features]
 kubernetes-discovery = [ "garage_rpc/kubernetes-discovery" ]
 k2v = [ "garage_util/k2v", "garage_api/k2v" ]
+# Exporter for the OpenTelemetry Collector.
+telemetry-otlp = [ "opentelemetry-otlp", "garage_api/telemetry-otlp" ]
diff --git a/src/garage/main.rs b/src/garage/main.rs
--- a/src/garage/main.rs
+++ b/src/garage/main.rs
@@ -8,6 +8,7 @@
 mod cli;
 mod repair;
 mod server;
+#[cfg(feature = "telemetry-otlp")]
 mod tracing_setup;
 
 use std::net::SocketAddr;
diff --git a/src/garage/server.rs b/src/garage/server.rs
--- a/src/garage/server.rs
+++ b/src/garage/server.rs
@@ -15,6 +15,7 @@
 use garage_api::k2v::api_server::K2VApiServer;
 
 use crate::admin::*;
+#[cfg(feature = "telemetry-otlp")]
 use crate::tracing_setup::*;
 
 async fn wait_from(mut chan: watch::Receiver<bool>) {
@@ -46,9 +47,15 @@
 	info!("Initializing Garage main data store...");
 	let garage = Garage::new(config.clone(), db, background);
 
-	info!("Initialize tracing...");
+	#[cfg(feature = "telemetry-otlp")]
 	if let Some(export_to) = config.admin.trace_sink {
+		info!("Initialize tracing...");
+
+		#[cfg(feature = "telemetry-otlp")]
 		init_tracing(&export_to, garage.system.id)?;
+
+		#[cfg(not(feature = "telemetry-otlp"))]
+		warn!("Garage was built without OTLP exporter, admin.trace_sink is ignored.");
 	}
 
 	info!("Initialize Admin API server and metrics collector...");
